### ISSUE TITLE
Add fail.format() to make it easier to use formatted strings with fail()

### ISSUE THREAD

**Issue by rickeylev** [bazelbuild/starlark#256]
### Description of the feature request:

Make it easier to produce nice looking error messages with `fail()`.

So we can write this:

```
fail.format(
    "Unable to find interpreter '{name}' for hub '{hub}' for version '{version}': " +
    "make sure a toolchain for version {version} was registered in module {module}",
    name = name, hub = hub, version=version, module=module
)
```

I'm not sure if java-backed starlark objects can have attribute lookup; a subattribute is appealing because it avoids introducing another global name and reads nicely.

Instead of this:

```
fail.format(
    (
        "Unable to find interpreter '{name}' for hub '{hub}' for version '{version}': " +
        "make sure a toolchain for version {version} was registered in module {module}"
    ).format(
        name=name, hub=hub, version=version, module=module
    )
)
```

You can play tricks with the location of newlines and parentheses to make the above look nicer, but it's tiring to fight the auto formatter for every fail() call.

There's several alternatives, but they are somewhat annoying:

* Create a msg variable, and do `fail(msg.format(...))`: This is annoying because assigning local vars just for a message is distracting. Because it's assigned to a local var, we start to think it's used elsewhere, but no, it just goes straight to fail.
* Lift the variable to a global: This moves the error text away from the local context it's occuring in. Now you have to jump back and forth to make sure your format-kwarg names are in the error message etc.
* Factor out a function: similar downside as a global var, but with more boilerplate of having to define a function that is called once and pass in all the local state the error message needs. Most fail-checks are localized and not re-used.

### What underlying problem are you trying to solve with this feature?

Making it easier to write useful informative error messages.

### Which operating system are you running Bazel on?

linux

### Have you found anything relevant by searching the web?

Didn't see anything

**Comment by ndmitchell** [comment]
This would be much more concise with f-strings #91, and in fact with that it probably is even better than the initial proposal.

```python
fail.format(
    f"Unable to find interpreter {name} for hub {hub} for version {version}: " +
    f"make sure a toolchain for version {version} was registered in module {module}"
)
```

**Comment by rickeylev** [comment]
A triple quoted string introduces extra leading white space. If it's pulled all the way to the left, then after a couple level of indentions, the flow of code is pretty awkward. Having to escape newlines to prevent them from becoming newlines is also awkward.

If there was a text.dedent function that took care of reflowing the text, that'd help.

A variation of fail.format would be to have fail accept varargs of message strings, and kwargs of format names, then internally concat and call format.

**Comment by brandjon** [comment]
I kinda like the possibility of adding f-strings to avoid `foo=foo` noise, but that doesn't help with the indentation problems of triple-quoted string literals.

I think a varargs of concatenated message strings goes against the convention of passing a string followed by positional or keyword substitution args. I think it's possibly confusing to violate that convention even if we're not supporting that style.

You'd think Python would've added a multiline string notation more similar to Java's text blocks, which don't suffer from the same indentation problems. It feels a little heavy handed, but you could imagine adding a `dedent()` method or similar to the string type. Not my first preference, but it's hard to think of an elegant and concise way to do this that isn't error-prone.

**Comment by brandjon** [comment]
> I think a varargs of concatenated message strings goes against the convention of passing a string followed by positional or keyword substitution args.

I say that, forgetting that `fail()` is already specified to format a sequence of *args, and in the Java implementation, supports an optional custom `sep` separator. So we should think of `fail()` as following the convention of `print()`.

For this FR, I don't think we need a format helper that's specific to `fail` or that should go underneath its "namepace" as `fail.format()`. Also note that `fail()`'s exact formatting of values is implementation-dependent, and in the Java interpreter it uses the value's debug representation, which is not generally safe for reifying back into the Starlark computation as a string value since it may contain non-deterministic information.

So I think this issue can be closed. The main request seems to be potentially addressed if we add f-strings (#91).
